internal/repository/prisma/dbsqlc/step_runs.sql

```diff
@@ -274,105 +275,186 @@ FROM step_run_data
 RETURNING *;
 
 -- name: ListStepRunsToReassign :many
-SELECT
-    sr.*
-FROM
-    "StepRun" sr
-LEFT JOIN
-    "Worker" w ON sr."workerId" = w."id"
-JOIN
-    "Step" s ON sr."stepId" = s."id"
+WITH valid_workers AS (
+    SELECT
+        DISTINCT ON (w."id")
+        w."id",
+        COALESCE(SUM(ws."slots"), 100) AS "slots"
+    FROM
+        "Worker" w
+    LEFT JOIN
+        "WorkerSemaphore" ws ON ws."workerId" = w."id"
+    WHERE
+        w."tenantId" = @tenantId::uuid
+        AND w."lastHeartbeatAt" > NOW() - INTERVAL '5 seconds'
+    GROUP BY
+        w."id"
+),
+-- Count the total number of slots across all workers
+total_max_runs AS (
+    SELECT
+        SUM("slots") AS "totalMaxRuns"
+    FROM
+        valid_workers
+),
+limit_max_runs AS (
+    SELECT
+        GREATEST("totalMaxRuns", 100) AS "limitMaxRuns"
+    FROM
+        total_max_runs
+),
+step_runs AS (
+    SELECT
+        sr.*
+    FROM
+        "StepRun" sr
+    LEFT JOIN
+        "Worker" w ON sr."workerId" = w."id"
+    JOIN
+        "JobRun" jr ON sr."jobRunId" = jr."id"
+    JOIN
+        "Step" s ON sr."stepId" = s."id"
+    WHERE
+        sr."tenantId" = @tenantId::uuid
+        AND ((
+            sr."status" = 'RUNNING'
+            AND w."lastHeartbeatAt" < NOW() - INTERVAL '60 seconds'
+            AND s."retries" > sr."retryCount"
+        ) OR (
+            sr."status" = 'ASSIGNED'
+            AND w."lastHeartbeatAt" < NOW() - INTERVAL '60 seconds'
+        ))
+        AND jr."status" = 'RUNNING'
+        AND sr."input" IS NOT NULL
+        -- Step run cannot have a failed parent
+        AND NOT EXISTS (
+            SELECT 1
+            FROM "_StepRunOrder" AS order_table
+            JOIN "StepRun" AS prev_sr ON order_table."A" = prev_sr."id"
+            WHERE 
+                order_table."B" = sr."id"
+                AND prev_sr."status" != 'SUCCEEDED'
+        )
+    ORDER BY
+        sr."createdAt" ASC
+    LIMIT
+        (SELECT "limitMaxRuns" FROM limit_max_runs)
+),
+locked_step_runs AS (
+    SELECT
+        sr."id", sr."status", sr."workerId"
+    FROM
+        step_runs sr
+    FOR UPDATE SKIP LOCKED
+)
+UPDATE
+    "StepRun"
+SET
+    "status" = 'PENDING_ASSIGNMENT',
+    -- requeue after now plus 4 seconds
+    "requeueAfter" = CURRENT_TIMESTAMP + INTERVAL '4 seconds',
```

Comment:
not really related to this PR but I'm wondering if we should make this configurable so we can adapt this in the load test to see if it has any influence on the general reliability and performance?

Reply:
Yep will file an issue